### PR TITLE
Add KomissarovShock analytic solution

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

This PR adds an analytic solution for the shock tests described in [Komissarov 1999](https://academic.oup.com/mnras/article-abstract/303/2/343/1067242), sec. 6 / Fig. 3. These tests were also done in the spectre 2017 paper. In a few preliminary runs comparing limiters I find the follow results in spectre:

bildschirmfoto 2018-12-05 um 18 49 10

The colours indicate the solution at different times t=0 to t=1.9.

- [x] Add test
- [x] Add documentation

### Types of changes:

- [x] New feature

### Component:

- [x] Code

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Comment by kidder** [comment]
We may need a different way of setting initial data for these shock problems.  Ideally we would like to have an inter-Block boundary where the data is discontinuous and set the values based on the BlockId instead of the coordinates.

**Comment by nilsvu** [comment]
@kidder let's talk about that at the Europe meeting on Monday

File: src/PointwiseFunctions/AnalyticData/GrMhd/KomissarovShock.hpp
Comment (mar-celine) [inline src/PointwiseFunctions/AnalyticData/GrMhd/KomissarovShock.hpp]: I would model the documentation here after what we have for BondiHoyleAccretion and CylindricalBlastWave :+1:

File: src/PointwiseFunctions/AnalyticData/GrMhd/KomissarovShock.hpp
Comment (mar-celine) [inline src/PointwiseFunctions/AnalyticData/GrMhd/KomissarovShock.hpp]: I would either change this so that it works for a general (but still physical) adiabatic index or add a comment as to why only a single value is used.

File: src/PointwiseFunctions/AnalyticData/GrMhd/KomissarovShock.cpp
Comment (mar-celine) [inline src/PointwiseFunctions/AnalyticData/GrMhd/KomissarovShock.cpp]: Is there a particular reason for using `inner_value` and `outer_value` as opposed to `left_value` and `right_value`? The current names make me think more of of a box function, which I don't think this is.

File: src/PointwiseFunctions/AnalyticData/GrMhd/KomissarovShock.cpp
Comment (mar-celine) [inline src/PointwiseFunctions/AnalyticData/GrMhd/KomissarovShock.cpp]: It looks to me that this is a shock along the x-direction only? Unless we know that we don't need a general-direction shock, it would be nice to make this a full-3D AnalyticData :smile: 

Edit: Regarding @kidder 's comment, it seems that the shock boundary will need to lie along an inter-block boundary, so I would defer to him in developing how the final shock profile looks in the AnalyticData.

**Review by mar-celine — COMMENTED** [review]
I notice that this has been marked in progress, so you are probably in the middle of adding your tests :+1:

**Comment by mar-celine** [comment]
I am removing the `review_help wanted` while the `in_progress` label is still up to avoid saturating the label/giving mixed messages, feel free to add it again if desired once `in_progress` no longer applies :octocat:

File: src/PointwiseFunctions/AnalyticData/GrMhd/KomissarovShock.cpp
Comment (nilsvu) [reply src/PointwiseFunctions/AnalyticData/GrMhd/KomissarovShock.cpp]: This is a 1D test, so we probably don't need general shock propagation directions. [Komissarov](https://academic.oup.com/mnras/article-abstract/303/2/343/1067242) does run another test where the propagation direction is rotated by 45deg, but not sure if we want to do that, too. It's not done in the spectre 2017 paper.

**Comment by nilsvu** [comment]
@gsb76 This was still quite unfinished, that's why it was tagged _in progress_. Should be ready for review now, though :)

**Comment by nilsdeppe** [comment]
Basically all of travis is unhappy with this. Marking as "in progress" for now until that's resolved

**Comment by nilsvu** [comment]
@nilsdeppe @gsb76 sorry I had just forgotten to push the changes 🤦‍♂️

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/KomissarovShock.cpp
Comment (fmahebert) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhd/KomissarovShock.cpp]: You should be able to use `step_function` to replace this `if ... else ...`, with the benefit that the same implementation should work for `double` and `DataVector` (I think). For an example, see the Burgers Step solution (although that one has left_value > right_value). Would need to check how `step_function`'s behavior at 0 matches what you are looking for.

Same comment for vector case below (presumably, one `step_function` per vector component)

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/KomissarovShock.cpp
Comment (fmahebert) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhd/KomissarovShock.cpp]: I don't think you gain anything by taking these `double`s by reference, which would lead me to prefer taking them by value for greater simplicity.

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/KomissarovShock.cpp
Comment (fmahebert) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhd/KomissarovShock.cpp]: Does `const double t` work here and in the similar below functions?

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/KomissarovShock.cpp
Comment (fmahebert) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhd/KomissarovShock.cpp]: I think this can use the existing function `dot_product(spatial_velocity, spatial_velocity)`

**Review by fmahebert — CHANGES_REQUESTED** [review]
Small things

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/KomissarovShock.cpp
Comment (nilsdeppe) [reply src/PointwiseFunctions/AnalyticSolutions/GrMhd/KomissarovShock.cpp]: Agreed, and often you lose something (performance) because the compiler ends up treating it as a pointer under the hood and you have more instructions (not many, but still :) )

Comment (nilsvu) [reply src/PointwiseFunctions/AnalyticSolutions/GrMhd/KomissarovShock.cpp]: Right, this was left over from when I tried to have a generic implementation for the scalar and vector function. Fixed.

**Comment by nilsvu** [comment]
fixed @fmahebert

**Comment by fmahebert** [comment]
I'm happy with the fixups, you can squash 👍. Travis, however, doesn't seem happy.

**Comment by nilsdeppe** [comment]
removed updated since, as @fmahebert pointed out, Travis is disappointed with this PR at the moment :P also needs a rebase. Please ping when you've squashed, rebased and appeased the Travis

**Comment by nilsvu** [comment]
ping @fmahebert @nilsdeppe

**Comment by nilsvu** [comment]
Travis is still angry, but not at me 😇

**Comment by nilsdeppe** [comment]
restarted segfault build, we can ignore the two index constraint one.

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/KomissarovShock.hpp
Comment (wthrowe) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhd/KomissarovShock.hpp]: `\ge`

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/KomissarovShock.hpp
Comment (nilsvu) [reply src/PointwiseFunctions/AnalyticSolutions/GrMhd/KomissarovShock.hpp]: fixed, but I guess this would be precisely the sort of review workflow we want to avoid with the streamlined review process we discussed earlier with @kidder @nilsdeppe 😄

**Review by wthrowe — DISMISSED** [review]

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/KomissarovShock.hpp
Comment (wthrowe) [reply src/PointwiseFunctions/AnalyticSolutions/GrMhd/KomissarovShock.hpp]: There have been several discussions, and I'm afraid I don't know which one you are talking about.  No one has ever bothered to propose guidelines or a review policy based on any of them, so you can't really expect people to follow ideas from whatever discussion you are thinking of.

In the lastest discussion I know of (which you were not part of) I was actually _encouraged_ to hold off on reviewing PRs until a substantial amount of review had happened.  So everyone has different expectations and I can't follow all of them.

**Review by fmahebert — DISMISSED** [review]

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/KomissarovShock.hpp
Comment (nilsdeppe) [reply src/PointwiseFunctions/AnalyticSolutions/GrMhd/KomissarovShock.hpp]: @nilsleiffischer @wthrowe is working on his thesis right now so Larry and I have intentionally left Will until the end so that the review is easier for him. Hopefully you can understand that.

I also agree with Will that such a discussion of PR review processes first needs to happen in the SpECTRE call and we need to have them in writing. I'm assuming that your (@nilsleiffischer) comment was merely meant to point something out to Larry and I. Let's not do that going forward or point it out offline so we don't end up with these sorts of possibly confusing conversations :)

Comment (nilsvu) [reply src/PointwiseFunctions/AnalyticSolutions/GrMhd/KomissarovShock.hpp]: I was indeed referring to the conversation we just had on the call and meant to adress @nilsdeppe and @kidder, I should have made that more clear @wthrowe. No criticism on your review intended

**Review by nilsdeppe — DISMISSED** [review]

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/KomissarovShock.hpp
Comment (kidder) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhd/KomissarovShock.hpp]: remove

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/KomissarovShock.hpp
Comment (wthrowe) [reply src/PointwiseFunctions/AnalyticSolutions/GrMhd/KomissarovShock.hpp]: @nilsdeppe please stop trying to explain to people why I'm doing things or what I'm talking about.  Your explanation wasn't even close to the mark in this case.

What I was refering to was the request that we allow a variety of people to do initial reviews of PRs so that they can gain more experience with the review process.  The more experienced reviwers were encouraged to hold off on doing detailed reviews if there was an active review going on.

**Review by kidder — CHANGES_REQUESTED** [review]
squash this

**Comment by kidder** [comment]
@nilsleiffischer clang-tidy is complaining

**Comment by nilsvu** [comment]
@kidder fixed (I think clang-tidy was wrong, but I don't feel like arguing with clang-tidy today :smile:)

**Comment by nilsdeppe** [comment]
You may disagree with the suggestion, but the statement it makes is correct according to the C++ standard :)

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/KomissarovShock.hpp
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhd/KomissarovShock.hpp:204]: This is where clang-tidy doesn't want me to have `const double t` because it says this is a declaration, not a definition. But isn't it just wrong and this is indeed a definition?

Comment (nilsdeppe) [reply src/PointwiseFunctions/AnalyticSolutions/GrMhd/KomissarovShock.hpp:204]: Ah, then yes. Sorry, I thought you meant you wanted the `const` in declarations :) In this case I'd do a `NOLINT`, but up to you :)

Comment (nilsvu) [reply src/PointwiseFunctions/AnalyticSolutions/GrMhd/KomissarovShock.hpp:204]: The other analytic solutions don't `const` this either, so I'll leave it as it is now.

**Review by fmahebert — APPROVED** [review]

**Comment by nilsvu** [comment]
I think this is done @sxs-collaboration/spectre-core-devs

**Review by kidder — APPROVED** [review]

**Pull request merged**
